File: android/core/domain/src/main/kotlin/com/cogra/domain/signing/WriteSigner.kt
// The session-authorized write orchestration: pre-sign → submit → (await
// seal) → verify → approve → observe (architecture.md "The write path").
// Handshake material persists in the IdentityStore between the two
// signatures so the flow survives process death without ever trusting
// the server's echo of the nonce; `resume` picks up every persisted
// handshake wherever it stopped.

package com.cogra.domain.signing

import com.cogra.crypto.ActorKey
import com.cogra.crypto.Family
import com.cogra.crypto.HandshakeException
import com.cogra.crypto.PreSignedProposal
import com.cogra.crypto.WireException
import com.cogra.domain.ErrorCode
import com.cogra.domain.Outcome
import com.cogra.domain.PreparedWriteView
import com.cogra.domain.StagedWriteView
import com.cogra.domain.UserError
import com.cogra.domain.WriteState
import com.cogra.domain.repo.WriteRepository
import com.cogra.domain.store.IdentityStore
import javax.inject.Inject
import kotlinx.coroutines.delay

/** Where one write's handshake ended up after a signing pass. */
sealed interface WriteResult {
    val stagedWriteId: String

    /** Approved and relaying (or already landed) — the device's part is done. */
    data class Done(override val stagedWriteId: String, val state: WriteState) : WriteResult

    /** Sealing hadn't returned within the polling budget; `resume` continues it. */
    data class AwaitingSeal(override val stagedWriteId: String) : WriteResult

    /** The API refused; the handshake is over. */
    data class Refused(override val stagedWriteId: String, val errors: List<UserError>) : WriteResult

    /** The device refused to sign — the sealed act failed verification. */
    data class RejectedByDevice(override val stagedWriteId: String, val reason: String) : WriteResult

    /** Transport failed mid-handshake; material is kept and `resume` retries. */
    data class Failed(override val stagedWriteId: String, val cause: Exception) : WriteResult
}

/** Thrown when signing is requested with no actor key on the device. */
class NoActorKeyException : Exception("no actor key on this device")

class WriteSigner @Inject constructor(
    private val writes: WriteRepository,
    private val identity: IdentityStore,
) {
    /** Seal-await polling budget: the stand-in seals synchronously, so a
     *  short bounded poll only covers the asynchronous contract. */
    private val sealPollAttempts = 5
    private val sealPollDelayMs = 1_000L

    /**
     * Runs the two-signature handshake for every prepared write. Each
     * write is its own priced act — one result per write, no cross-write
     * atomicity (api-spec.md "The write flow").
     */
    suspend fun sign(prepared: List<PreparedWriteView>): List<WriteResult> {
        val key = actorKey()
        return prepared.map { write -> signOne(key, write) }
    }

    /**
     * Signs one staged write served mid-handshake — the path of the
     * staged Registration, which the backend stages at approval and the
     * device discovers on the status poll (auth.md "Approval and
     * landing"): same two signatures, same verification, through the
     * ordinary session-authorized legs.
     */
    suspend fun signStaged(staged: StagedWriteView): WriteResult {
        if (staged.family == Family.UNKNOWN) return unsupported(staged.id, "record family")
        val key = actorKey()
        return when (staged.state) {
            WriteState.AWAITING_PRE_SIGN -> when (val pre = identity.handshake(staged.id)) {
                // A fresh staging: pre-sign from the served proposal.
                null -> {
                    val step = try {
                        preSignStep(key, staged.canonicalProposal)
                    } catch (e: WireException) {
                        return WriteResult.RejectedByDevice(staged.id, e.message ?: "malformed proposal")
                    }
                    identity.saveHandshake(staged.id, step.pre)
                    when (val submitted = writes.submitProposal(staged.id, step.signatureBase64)) {
                        is Outcome.Success -> approveFrom(key, step.pre, submitted.value)
                        is Outcome.Refused -> refused(staged.id, submitted.errors)
                        is Outcome.Failed -> WriteResult.Failed(staged.id, submitted.cause)
                    }
                }
                // The submit response was lost; re-send THIS device's material.
                else -> when (val submitted = writes.submitProposal(staged.id, preCommitmentSignature(pre))) {
                    is Outcome.Success -> approveFrom(key, pre, submitted.value)
                    is Outcome.Refused -> refused(staged.id, submitted.errors)
                    is Outcome.Failed -> WriteResult.Failed(staged.id, submitted.cause)
                }
            }
            WriteState.SEALING, WriteState.AWAITING_APPROVAL -> when (val pre = identity.handshake(staged.id)) {
                // Pre-signed by a device that is not this one (or the
                // material is gone): only the expiry re-stage recovers.
                null -> WriteResult.Refused(
                    staged.id,
                    listOf(UserError(ErrorCode.INTERNAL, "handshake material lost — awaiting re-stage")),
                )
                else -> approveFrom(key, pre, staged)
            }
            WriteState.RELAYING, WriteState.LANDED -> {
                identity.clearHandshake(staged.id)
                WriteResult.Done(staged.id, staged.state)
            }
            WriteState.EXPIRED -> {
                identity.clearHandshake(staged.id)
                WriteResult.Refused(
                    staged.id,
                    listOf(UserError(ErrorCode.STAGED_WRITE_EXPIRED, "garbage-collected unlanded")),
                )
            }
            WriteState.UNKNOWN -> unsupported(staged.id, "staged-write state")
        }
    }

    /**
     * Continues every persisted handshake from its server-side state —
     * the process-death and lost-response recovery path.
     */
    suspend fun resume(): List<WriteResult> {
        val ids = identity.handshakeIds()
        if (ids.isEmpty()) return emptyList()
        val key = actorKey()
        return ids.map { id -> resumeOne(key, id) }
    }

    private suspend fun actorKey(): ActorKey =
        ActorKey.fromSeed(identity.actorSeed() ?: throw NoActorKeyException())

    private suspend fun signOne(key: ActorKey, write: PreparedWriteView): WriteResult {
        if (write.family == Family.UNKNOWN) return unsupported(write.id, "record family")
        val step = try {
            preSignStep(key, write.canonicalProposal)
        } catch (e: WireException) {
            return WriteResult.RejectedByDevice(write.id, e.message ?: "malformed proposal")
        }
        identity.saveHandshake(write.id, step.pre)
        return when (val submitted = writes.submitProposal(write.id, step.signatureBase64)) {
            is Outcome.Success -> approveFrom(key, step.pre, submitted.value)
            is Outcome.Refused -> refused(write.id, submitted.errors)
            is Outcome.Failed -> WriteResult.Failed(write.id, submitted.cause)
        }
    }

    private suspend fun resumeOne(key: ActorKey, id: String): WriteResult {
        val pre = identity.handshake(id) ?: return WriteResult.Failed(id, IllegalStateException("no material"))
        return when (val read = writes.stagedWrite(id)) {
            is Outcome.Failed -> WriteResult.Failed(id, read.cause)
            is Outcome.Refused -> refused(id, read.errors)
            is Outcome.Success -> when (val staged = read.value) {
                // Gone entirely — collected long ago; nothing to continue.
                null -> {
                    identity.clearHandshake(id)
                    WriteResult.Refused(id, listOf(UserError(ErrorCode.NOT_FOUND, "staged write is gone")))
                }
                else -> if (staged.family == Family.UNKNOWN) {
                    unsupported(id, "record family")
                } else when (staged.state) {
                    // The submit response was lost before the server saw it.
                    WriteState.AWAITING_PRE_SIGN -> {
                        val sig = preCommitmentSignature(pre)
                        when (val submitted = writes.submitProposal(id, sig)) {
                            is Outcome.Success -> approveFrom(key, pre, submitted.value)
                            is Outcome.Refused -> refused(id, submitted.errors)
                            is Outcome.Failed -> WriteResult.Failed(id, submitted.cause)
                        }
                    }
                    WriteState.SEALING, WriteState.AWAITING_APPROVAL ->
                        approveFrom(key, pre, staged)
                    WriteState.RELAYING, WriteState.LANDED -> {
                        identity.clearHandshake(id)
                        WriteResult.Done(id, staged.state)
                    }
                    WriteState.EXPIRED -> {
                        identity.clearHandshake(id)
                        WriteResult.Refused(
                            id,
                            listOf(UserError(ErrorCode.STAGED_WRITE_EXPIRED, "garbage-collected unlanded")),
                        )
                    }
                    WriteState.UNKNOWN -> unsupported(id, "staged-write state")
                }
            }
        }
    }

    /** From a staged write in hand: await the seal if needed, verify, approve. */
    private suspend fun approveFrom(key: ActorKey, pre: PreSignedProposal, staged: StagedWriteView): WriteResult {
        var current = staged
        var attempts = 0
        while (current.verifiedAct == null) {
            if (current.state == WriteState.UNKNOWN) return unsupported(current.id, "staged-write state")
            if (current.state == WriteState.EXPIRED) {
                identity.clearHandshake(current.id)
                return WriteResult.Refused(
                    current.id,
                    listOf(UserError(ErrorCode.STAGED_WRITE_EXPIRED, "garbage-collected unlanded")),
                )
            }
            if (attempts++ >= sealPollAttempts) return WriteResult.AwaitingSeal(current.id)
            delay(sealPollDelayMs)
            current = when (val read = writes.stagedWrite(current.id)) {
                is Outcome.Success -> read.value ?: return WriteResult.Failed(
                    current.id,
                    IllegalStateException("staged write vanished mid-seal"),
                )
                is Outcome.Refused -> return refused(current.id, read.errors)
                is Outcome.Failed -> return WriteResult.Failed(current.id, read.cause)
            }
        }
        val host = when (val read = writes.hostPublicKey()) {
            is Outcome.Success -> read.value
            is Outcome.Refused -> return refused(current.id, read.errors)
            is Outcome.Failed -> return WriteResult.Failed(current.id, read.cause)
        }
        val witnessSignature = try {
            approveStep(key, pre, checkNotNull(current.verifiedAct), host)
        } catch (e: HandshakeException) {
            // The device refuses to sign; the material is spent — the
            // unapproved staging garbage-collects server-side.
            identity.clearHandshake(current.id)
            return WriteResult.RejectedByDevice(current.id, e.message ?: "verification failed")
        } catch (e: WireException) {
            identity.clearHandshake(current.id)
            return WriteResult.RejectedByDevice(current.id, e.message ?: "malformed verified act")
        }
        return when (val approved = writes.approveAct(current.id, witnessSignature)) {
            is Outcome.Success -> {
                // The server holds the witness; the device's part is done.
                identity.clearHandshake(current.id)
                WriteResult.Done(current.id, approved.value.state)
            }
            is Outcome.Refused -> refused(current.id, approved.errors)
            is Outcome.Failed -> WriteResult.Failed(current.id, approved.cause)
        }
    }

    /** A refusal ends the handshake; spent material never resumes. */
    private suspend fun refused(id: String, errors: List<UserError>): WriteResult {
        identity.clearHandshake(id)
        return WriteResult.Refused(id, errors)
    }

    /**
     * A server value this app version does not know — refused WITHOUT
     * clearing the handshake material, so an updated build can resume.
     */
    private fun unsupported(id: String, what: String): WriteResult = WriteResult.Refused(
        id,
        listOf(UserError(ErrorCode.UNKNOWN, "$what this app version does not know — update the app")),
    )

    private fun preCommitmentSignature(pre: PreSignedProposal): String =
        java.util.Base64.getEncoder().encodeToString(com.cogra.crypto.encodePreCommitmentOf(pre))
}

File: web/src/lib/signing/write-signer.ts
// The device's half of the write handshake (api-spec.md "The write
// flow"; Android mirror: WriteSigner.kt): material persists BEFORE the
// submit, and the approve step verifies against that persisted copy,
// never the server's echo — the user never signs blind bytes.

import type { ApolloClient } from "@apollo/client";

import type { StagedWriteState } from "@/__generated__/graphql";
import type { UserError } from "@/lib/api/outcome";
import {
  approveAct,
  fetchStagedWrite,
  hostPublicKey,
  submitProposal,
  type StagedWriteView,
} from "@/lib/api/writes-api";
import { ActorKey, HandshakeError } from "@/lib/crypto/actor-key";
import { fromBase64, toBase64 } from "@/lib/crypto/bytes";
import type { PreSignedProposal } from "@/lib/crypto/handshake";
import { decodeProposal, decodeVerifiedAct, encodePreCommitmentOf, WireError } from "@/lib/crypto/wire";
import type { AuthGuard } from "@/lib/session/guard";
import type { IdentityStore } from "@/lib/identity/store";

export type WriteResult =
  /** Approved and relaying (or already landed) — the device's part is done. */
  | { kind: "done"; id: string; state: StagedWriteState }
  /** Seal-poll budget exhausted; material kept, resume() continues. */
  | { kind: "awaitingSeal"; id: string }
  /** The API refused; TERMINAL_REFUSALS decides whether material is cleared. */
  | { kind: "refused"; id: string; errors: readonly UserError[] }
  /** The device refused to sign a failing verification; material cleared. */
  | { kind: "rejectedByDevice"; id: string; reason: string }
  /** Transport fault; material kept, resume() retries. */
  | { kind: "failed"; id: string; cause: unknown };

export type WriteSigner = {
  /**
   * Runs the handshake for a staged write discovered on the poll — the
   * admission Registration's path. Dispatches on the staged state so a
   * crash between any two legs heals here.
   */
  signStaged(staged: StagedWriteView): Promise<WriteResult>;
  /** Continues every persisted handshake — the parked-acts entry point. */
  resume(): Promise<WriteResult[]>;
};

const SEAL_POLL_ATTEMPTS = 5;
const SEAL_POLL_DELAY_MS = 1_000;

/**
 * Refusal codes that end a handshake for good — the material is spent.
 * Anything else (RATE_LIMITED backoff, INTERNAL fault, UNAUTHENTICATED,
 * a code this build does not know) keeps the material for resume().
 */
const TERMINAL_REFUSALS: ReadonlySet<UserError["code"]> = new Set([
  "SIGNATURE_INVALID",
  "STAGED_WRITE_EXPIRED",
  "NOT_FOUND",
  "BAD_INPUT",
  "FORBIDDEN",
  "WRITE_RULE_FAILED",
]);

function synthesized(code: UserError["code"], message: string): UserError {
  return { code, message, field: null };
}

export function createWriteSigner(deps: {
  client: ApolloClient;
  guard: AuthGuard;
  store: IdentityStore;
  /** Injectable for tests; production uses the real timer. */
  sleep?: (ms: number) => Promise<void>;
}): WriteSigner {
  const { client, guard, store } = deps;
  const sleep = deps.sleep ?? ((ms) => new Promise((resolve) => setTimeout(resolve, ms)));

  /** A write-leg refusal: clears the material only when it is spent. */
  async function refuse(id: string, errors: readonly UserError[]): Promise<WriteResult> {
    if (errors.length > 0 && errors.every((e) => TERMINAL_REFUSALS.has(e.code))) {
      await store.clearHandshake(id);
    }
    return { kind: "refused", id, errors };
  }

  /** A read-leg refusal: the handshake is intact server-side — keep the material. */
  function refuseKeeping(id: string, errors: readonly UserError[]): WriteResult {
    return { kind: "refused", id, errors };
  }

  async function reject(id: string, reason: string): Promise<WriteResult> {
    await store.clearHandshake(id);
    return { kind: "rejectedByDevice", id, reason };
  }

  async function done(id: string, state: StagedWriteState): Promise<WriteResult> {
    await store.clearHandshake(id);
    return { kind: "done", id, state };
  }

  function preCommitmentSignature(pre: PreSignedProposal): string {
    return toBase64(encodePreCommitmentOf(pre));
  }

  /**
   * From a submitted pre-commitment to the approval: await the seal
   * (bounded — the seal may be synchronous, in which case the loop never
   * runs), verify everything, sign the witness.
   */
  async function approveFrom(
    key: ActorKey,
    id: string,
    pre: PreSignedProposal,
    staged: StagedWriteView,
  ): Promise<WriteResult> {
    let current = staged;
    let attempts = 0;
    while (current.verifiedAct === null) {
      if (current.state === "EXPIRED") {
        return refuse(id, [synthesized("STAGED_WRITE_EXPIRED", "garbage-collected unlanded")]);
      }
      if (attempts >= SEAL_POLL_ATTEMPTS) return { kind: "awaitingSeal", id };
      attempts += 1;
      await sleep(SEAL_POLL_DELAY_MS);
      const read = await guard.run(() => fetchStagedWrite(client, id));
      if (read.kind === "refused") return refuseKeeping(id, read.errors);
      if (read.kind === "failed") return { kind: "failed", id, cause: read.cause };
      if (read.value === null) {
        return { kind: "failed", id, cause: new Error("staged write vanished mid-seal") };
      }
      current = read.value;
    }

    const hostKey = await hostPublicKey(client);
    if (hostKey.kind === "refused") return refuseKeeping(id, hostKey.errors);
    if (hostKey.kind === "failed") return { kind: "failed", id, cause: hostKey.cause };

    let approvalSignature: string;
    try {
      const act = decodeVerifiedAct(fromBase64(current.verifiedAct));
      const witness = await key.approve(pre, act, fromBase64(hostKey.value));
      approvalSignature = toBase64(witness.approvalSignature);
    } catch (e) {
      if (e instanceof HandshakeError || e instanceof WireError) {
        return reject(id, e.message);
      }
      throw e;
    }

    const approved = await guard.run(() => approveAct(client, id, approvalSignature));
    if (approved.kind === "refused") return refuse(id, approved.errors);
    if (approved.kind === "failed") return { kind: "failed", id, cause: approved.cause };
    return done(id, approved.value.state);
  }

  async function submitAndApprove(
    key: ActorKey,
    id: string,
    pre: PreSignedProposal,
  ): Promise<WriteResult> {
    const submitted = await guard.run(() => submitProposal(client, id, preCommitmentSignature(pre)));
    if (submitted.kind === "refused") return refuse(id, submitted.errors);
    if (submitted.kind === "failed") return { kind: "failed", id, cause: submitted.cause };
    return approveFrom(key, id, pre, submitted.value);
  }

  async function signOne(key: ActorKey, staged: StagedWriteView): Promise<WriteResult> {
    const id = staged.id;
    const pre = await store.handshake(id);
    switch (staged.state) {
      case "AWAITING_PRE_SIGN": {
        if (pre !== null) {
          // The submit response was lost: re-send THIS device's material
          // — the stored nonce, never a fresh one.
          return submitAndApprove(key, id, pre);
        }
        let fresh: PreSignedProposal;
        try {
          fresh = await key.preSign(decodeProposal(fromBase64(staged.canonicalProposal)));
        } catch (e) {
          if (e instanceof WireError) return reject(id, e.message);
          throw e;
        }
        await store.saveHandshake(id, fresh);
        return submitAndApprove(key, id, fresh);
      }
      case "SEALING":
      case "AWAITING_APPROVAL": {
        if (pre !== null) return approveFrom(key, id, pre, staged);
        // Pre-signed by another device, or custody was cleared. Not a
        // device rejection — only the expiry re-stage recovers.
        return refuse(id, [
          synthesized("INTERNAL", "handshake material lost — awaiting re-stage"),
        ]);
      }
      case "RELAYING":
      case "LANDED":
        return done(id, staged.state);
      case "EXPIRED":
        return refuse(id, [synthesized("STAGED_WRITE_EXPIRED", "garbage-collected unlanded")]);
      default:
        // A state this build does not know (the generated union lags the
        // server): refuse without spending the material — an updated
        // build resumes it. Android's WriteState.UNKNOWN twin.
        return refuseKeeping(id, [
          synthesized("INTERNAL", "staged-write state this app version does not know"),
        ]);
    }
  }

  async function resumeOne(key: ActorKey, id: string): Promise<WriteResult> {
    const pre = await store.handshake(id);
    if (pre === null) {
      return { kind: "failed", id, cause: new Error("no handshake material for this id") };
    }
    const read = await guard.run(() => fetchStagedWrite(client, id));
    if (read.kind === "refused") return refuseKeeping(id, read.errors);
    if (read.kind === "failed") return { kind: "failed", id, cause: read.cause };
    if (read.value === null) {
      // Collected long ago — the handshake can never finish.
      return refuse(id, [synthesized("NOT_FOUND", "staged write is gone")]);
    }
    return signOne(key, read.value);
  }

  return {
    async signStaged(staged) {
      const key = await store.actorKey();
      if (key === null) {
        return { kind: "failed", id: staged.id, cause: new Error("no actor key on this device") };
      }
      return signOne(key, staged);
    },

    async resume() {
      const ids = await store.handshakeIds();
      if (ids.length === 0) return [];
      const key = await store.actorKey();
      if (key === null) {
        return ids.map((id) => ({
          kind: "failed" as const,
          id,
          cause: new Error("no actor key on this device"),
        }));
      }
      const results: WriteResult[] = [];
      for (const id of ids) results.push(await resumeOne(key, id));
      return results;
    },
  };
}
